### ISSUE TITLE
encode raw condition as output's condition

### ISSUE THREAD

**Issue by GlenDC** [rivine/rivine#259]
I propose to start using the output condition directly, instead of the unlock hash (which is generated using the output condition anyhow), as part of a coin/blockStake output.

This would require a new transaction version, one that would look as:

```go
TransactionV1 struct { // 0x01, as a replacement for the current and original transaction version 0x00
    CoinInputs        []struct{
        ParentID    CoinOutputID           `json:"parentid"`
        Fulfillment UnlockFulfillmentProxy `json:"fulfillment"`
    }
    CoinOutputs       []struct{
        Value     Currency             `json:"value"`
        Condition UnlockConditionProxy `json:"condition"`
    }
    BlockStakeInputs  []struct{
        ParentID    BlockstakeOutputID     `json:"parentid"`
        Fulfillment UnlockFulfillmentProxy `json:"fulfillment"`
    }
    BlockStakeOutputs []struct{
        Value     Currency             `json:"value"`
        Condition UnlockConditionProxy `json:"condition"`
    }
    MinerFees     []Currency
    ArbitraryData []byte

    _VersionNumber TransactionVersion // 0x01
}
```

This would make a lot of things easier:

+ Wallets have all the info they would ever need to decide if an output (using whatever known unlock condition) is destined for them, and if so, how to handle/bucket/categorize/use that output;
+ When claiming/refunding an atomic swap contract you would no longer need to give the condition, as in, that could be optional, as the condition could be taken from the given outputID;

Existing transactions with the original transaction version (`0x00`, let's call it `TransactionV0`), would still be supported, but we would no longer create them ourselves (even though we safely could).

This should not bring any security risks with it, as the condition was anyhow already considered public information, even though it isn't now as easily visible as it would be now. Now it would be actually properly visible, just as PubSigKeys of bitcoin are visible as well, in their raw form.

Extra notes about these changes:

+ While we would keep the InputLock stuff and all that code in as legacy reasons (as to keep legacy support for v0 transactions), we would now use the new separated Condition/Fulfillment (proxy) code, which will still need to be implemented, as part of v1 transactions;
+ Coin/BlockStake inputs would no longer define the condition itself, and only give the fulfillment. As the referenced output anyhow has to exist on the blockchain prior to this input, we can guarantee that we'll always be able to get the condition from the referenced output, as part of the validation process;

Open question: Should as part of the validation the unlock type of the input and referenced output always be equal? Quickly thinking I would say yes,  but what if in the future we want to allow multiple types of fulfillments for a single (specific type of) condition? I think in general that doesn't make any sense and these unlock types should always be equal, but I leave it as an open question here. My personal opinion is thus that the unlock type of the input and referenced output should be equal, but if someone else has another opinion about this, now is the time to share it.

**Comment by GlenDC** [comment]
The javascript example given in https://github.com/rivine/rivine/blob/master/doc/transactions/transaction.md#text-format, would , using this new proposed transaction version (`0x01`), become:

```javascript
{
	"version": 1, // version ID, byte, new version `0x01`,
	"data": { // actual transaction, required
		"coininputs": [ // coin inputs, at least 1 element required
			{
				// coin output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
				"parentid": "abcdef012345abcdef012345abcdef012345abcdef012345abcdef012345abcd",
				"unlocktype": 1,
				// unlock fulfillment, fulfills the unlock condition, required,
				// format dependend upon sibling "unlocktype" property
				"fulfillment": {
					// signature of the transaction using the private key,
					// linked to the public key given as condition of the referenced output.
					"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab"
				},
			},
			{
				// coin output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
				"parentid": "012345defabc012345defabc012345defabc012345defabc012345defabc0123",
				"unlocktype": 2,
				"fulfillment": { 	// unlock fulfillment, fulfills the unlock condition, required,
									// format dependend upon sibling "type" property
					// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
					// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
					// <algorithmSpecifier> can currently only be `"ed25519"`
					"publickey": "ed25519:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
					// signature, byte-size is fixed but dependend upon the <algorithmSpecifier>,
					// when <algorithmSpecifier> equals `"ed25519"` the byte size is 64, required, hex-encoded
					"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab",
					// secret, fixed size, 32 bytes, hex-encoded
					"secret": "def789def789def789def789def789dedef789def789def789def789def789de"
				}
			}
		],
		"coinoutputs": [ // coin outputs, optional
			{
				"value": "3", // currency value (in smallest unit), big integer as string, required, positive
				// output's condition, the destined has to fulfill in order to use this output as an input
				"unlocktype": 1,
				"condition": {
					"publickey": "ed25519:eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
				}
			},
			{
				"value": "5", // currency value (in smallest unit), big integer as string, required, positive
				// output's condition, the destined has to fulfill in order to use this output as an input
				"unlocktype": 1,
				"condition": {
					"publickey": "ed25519:defdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefa"
				}
			},
			{
				"value": "8", // currency value (in smallest unit), big integer as string, required, positive
				// output's condition, the destined has to fulfill in order to use this output as an input
				"unlocktype": 2,
				"condition": {
					// sender's unlock hash, required, hex-encoded, fixed-size
					"sender": "0101234567890123456789012345678901012345678901234567890123456789018a50e31447b8",
					// receiver's unlock hash, required, hex-encoded, fixed-size
					"receiver": "01abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc057382370c8d9",
					// hashed secret, fixed size, 32 bytes, hex-encoded, sha256(secret)
					"hashedsecret": "abc543defabc543defabc543defabc543defabc543defabc543defabc543defa",
					// time lock, unix epoch timestamp (in seconds), 64-bit unsigned integer, required
					"timelock": 1522068743
				}
			}
		],
		"blockstakeinputs": [ // block stake inputs, optional
			{
				// blockstake output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
				"parentid": "dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfde",
				"unlocktype": 1,
				// unlock fulfillment, fulfills the unlock condition, required,
				// format dependend upon sibling "unlocktype" property
				"fulfillment": {
					// signature of the transaction using the private key,
					// linked to the public key given as condition of the referenced output.
					"signature": "01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def"
				}
			},
			{
				// blockstake output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
				"parentid": "fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed4",
				"unlocktype": 42, // unknwon type example
				// unlock fulfillment, fulfills the unlock condition, required,
				// format dependend upon sibling "unlocktype" property
				"fulfillment": "ZnVsZmlsbG1lbnQ=" // base64 encoding of binary encoded fulfillment is used for unknown types
			}
		],
		"blockstakeoutputs": [ // block stake outputs, optional
			{
				"value": "4", // currency value (in smallest unit), big integer as string, required, positive
				"unlocktype": 1,
				"condition": {
					"publickey": "ed25519:defdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefa"
				}
			},
			{
				"value": "2", // currency value (in smallest unit), big integer as string, required, positive
				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
				"unlocktype": 42, // unknown type
				"condition": "Y29uZGl0aW9u" // base64 encoding of binary encoded condition is used for unknown types
			}
		],
		// miner fees, list of currency values, at least 1 required
		"minerfees": ["1", "2", "3"],
		// arbitrary data, optional, base64 encoded
		"arbitrarydata": "ZGF0YQ==" // represents "data" in base64 encoding
	}
}
```

To give a quick overview with what has changed compared to the previous (original transaction version):

```diff
--- /Users/glendc/tmp/transaction_v0.diff	2018-04-06 13:33:51.000000000 +0200
+++ /Users/glendc/tmp/transaction_v1.diff	2018-04-06 13:33:34.000000000 +0200
@@ -1,125 +1,105 @@
 {
-	"version": 0, // version ID, byte, for now always `0`, optional
+	"version": 1, // version ID, byte, new version `0x01`,
 	"data": { // actual transaction, required
 		"coininputs": [ // coin inputs, at least 1 element required
 			{
 				// coin output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
 				"parentid": "abcdef012345abcdef012345abcdef012345abcdef012345abcdef012345abcd",
-				// input lock, required
-				"unlocker": {
-					"type": 1, 	// input lock, byte, supported range: [1,2], unknown range: [3,255], required
-											// `1` = SingleSignature, `2` = AtomicSwap
-					"condition": { // unlock condition, required, format dependend upon sibling "type" property
-						// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
-						// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// <algorithmSpecifier> can currently only be `"ed25519"`
-						"publickey": "ed25519:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
-					},
-					"fulfillment": { 	// unlock fulfillment, fulfills the unlock condition, required,
-														// format dependend upon sibling "type" property
-						// signature, byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// when <algorithmSpecifier> equals `"ed25519"` the byte size is 64, required, hex-encoded
-						"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab"
-					}
-				}
+				"unlocktype": 1,
+				// unlock fulfillment, fulfills the unlock condition, required,
+				// format dependend upon sibling "unlocktype" property
+				"fulfillment": {
+					// signature of the transaction using the private key,
+					// linked to the public key given as condition of the referenced output.
+					"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab"
+				},
 			},
 			{
 				// coin output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
 				"parentid": "012345defabc012345defabc012345defabc012345defabc012345defabc0123",
-				// input lock, required
-				"unlocker": {
-					"type": 2, 	// input lock, byte, supported range: [1,2], unknown range: [3,255], required
-											// `1` = SingleSignature, `2` = AtomicSwap
-					"condition": { // unlock condition, required, format dependend upon sibling "type" property
-						// sender's unlock hash, required, hex-encoded, fixed-size
-						"sender": "0101234567890123456789012345678901012345678901234567890123456789018a50e31447b8",
-						// receiver's unlock hash, required, hex-encoded, fixed-size
-						"receiver": "01abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc057382370c8d9",
-						// hashed secret, fixed size, 32 bytes, hex-encoded, sha256(secret)
-						"hashedsecret": "abc543defabc543defabc543defabc543defabc543defabc543defabc543defa",
-						// time lock, unix epoch timestamp (in seconds), 64-bit unsigned integer, required
-						"timelock": 1522068743
-					},
-					"fulfillment": { 	// unlock fulfillment, fulfills the unlock condition, required,
-														// format dependend upon sibling "type" property
-						// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
-						// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// <algorithmSpecifier> can currently only be `"ed25519"`
-						"publickey": "ed25519:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
-						// signature, byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// when <algorithmSpecifier> equals `"ed25519"` the byte size is 64, required, hex-encoded
-						"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab",
-						// secret, fixed size, 32 bytes, hex-encoded
-						"secret": "def789def789def789def789def789dedef789def789def789def789def789de"
-					}
+				"unlocktype": 2,
+				"fulfillment": { 	// unlock fulfillment, fulfills the unlock condition, required,
+									// format dependend upon sibling "type" property
+					// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
+					// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
+					// <algorithmSpecifier> can currently only be `"ed25519"`
+					"publickey": "ed25519:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
+					// signature, byte-size is fixed but dependend upon the <algorithmSpecifier>,
+					// when <algorithmSpecifier> equals `"ed25519"` the byte size is 64, required, hex-encoded
+					"signature": "abcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefabcdefab",
+					// secret, fixed size, 32 bytes, hex-encoded
+					"secret": "def789def789def789def789def789dedef789def789def789def789def789de"
 				}
 			}
 		],
 		"coinoutputs": [ // coin outputs, optional
 			{
 				"value": "3", // currency value (in smallest unit), big integer as string, required, positive
-				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
-				"unlockhash": "0101234567890123456789012345678901012345678901234567890123456789018a50e31447b8"
+				// output's condition, the destined has to fulfill in order to use this output as an input
+				"unlocktype": 1,
+				"condition": {
+					"publickey": "ed25519:eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"
+				}
 			},
 			{
 				"value": "5", // currency value (in smallest unit), big integer as string, required, positive
-				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
-				"unlockhash": "01abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc057382370c8d9"
+				// output's condition, the destined has to fulfill in order to use this output as an input
+				"unlocktype": 1,
+				"condition": {
+					"publickey": "ed25519:defdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefa"
+				}
 			},
 			{
 				"value": "8", // currency value (in smallest unit), big integer as string, required, positive
-				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
-				"unlockhash": "02abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc0ee7715ac0e0e"
+				// output's condition, the destined has to fulfill in order to use this output as an input
+				"unlocktype": 2,
+				"condition": {
+					// sender's unlock hash, required, hex-encoded, fixed-size
+					"sender": "0101234567890123456789012345678901012345678901234567890123456789018a50e31447b8",
+					// receiver's unlock hash, required, hex-encoded, fixed-size
+					"receiver": "01abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc057382370c8d9",
+					// hashed secret, fixed size, 32 bytes, hex-encoded, sha256(secret)
+					"hashedsecret": "abc543defabc543defabc543defabc543defabc543defabc543defabc543defa",
+					// time lock, unix epoch timestamp (in seconds), 64-bit unsigned integer, required
+					"timelock": 1522068743
+				}
 			}
 		],
 		"blockstakeinputs": [ // block stake inputs, optional
 			{
 				// blockstake output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
 				"parentid": "dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfd23dfde",
-				// input lock, required
-				"unlocker": {
-					"type": 1, 	// input lock, byte, supported range: [1,2], unknown range: [3,255], required
-											// `1` = SingleSignature, `2` = AtomicSwap
-					"condition": { // unlock condition, required, format dependend upon sibling "type" property
-						// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
-						// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// <algorithmSpecifier> can currently only be `"ed25519"`
-						"publickey": "ed25519:ef1234ef1234ef1234ef1234ef1234ef1234ef1234ef1234ef1234ef1234ef12"
-					},
-					"fulfillment": { 	// unlock fulfillment, fulfills the unlock condition, required,
-														// format dependend upon sibling "type" property
-						// public key, required, format: `<algorithmSpecifier>:<key>`, where <key> is hex-encoded
-						// and which byte-size is fixed but dependend upon the <algorithmSpecifier>,
-						// <algorithmSpecifier> can currently only be `"ed25519"`
-						"signature": "01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def"
-					}
+				"unlocktype": 1,
+				// unlock fulfillment, fulfills the unlock condition, required,
+				// format dependend upon sibling "unlocktype" property
+				"fulfillment": {
+					// signature of the transaction using the private key,
+					// linked to the public key given as condition of the referenced output.
+					"signature": "01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def01234def"
 				}
 			},
 			{
 				// blockstake output ID, crypto hash (blake2b, 256 bit), required, hex-encoded
 				"parentid": "fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed42fed4",
-				"unlocker": {
-					"type": 42, // input lock, byte, supported range: [1,2], unknown range: [3,255], required
-											// `1` = SingleSignature, `2` = AtomicSwap, other = unknown
-					// unlock condition, required,
-					// base64 encoding of the input lock condition in binary encoding format,
-					"condition": "Y29uZGl0aW9u",
-					// unlock fulfillment, required,
-					// base64 encoding of the input fulfillment condition in binary encoding format,
-					"fulfillment": "ZnVsZmlsbG1lbnQ="
-				}
+				"unlocktype": 42, // unknwon type example
+				// unlock fulfillment, fulfills the unlock condition, required,
+				// format dependend upon sibling "unlocktype" property
+				"fulfillment": "ZnVsZmlsbG1lbnQ=" // base64 encoding of binary encoded fulfillment is used for unknown types
 			}
 		],
 		"blockstakeoutputs": [ // block stake outputs, optional
 			{
 				"value": "4", // currency value (in smallest unit), big integer as string, required, positive
-				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
-				"unlockhash": "2a0123456789012345678901234567890101234567890123456789012345678901baa780c3d6f2"
+				"unlocktype": 1,
+				"condition": {
+					"publickey": "ed25519:defdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefdefa"
+				}
 			},
 			{
 				"value": "2", // currency value (in smallest unit), big integer as string, required, positive
 				// output's unlock hash, specifier output will belocked to that hash, could represent wallet
-				"unlockhash": "18abc0123abc0123abc0123abc0123abc0abc0123abc0123abc0123abc0123abc09c8b0a79cf2b"
+				"unlocktype": 42, // unknown type
+				"condition": "Y29uZGl0aW9u" // base64 encoding of binary encoded condition is used for unknown types
 			}
 		],
 		// miner fees, list of currency values, at least 1 required
```

**Comment by GlenDC** [comment]
Blocked by #201.

**Comment by GlenDC** [comment]
Resolved by 5186d739e7d6001e6c060bc33eb341b22d6478c8.
